### PR TITLE
crimson/net: improve batching in the write path

### PR DESCRIPTION
**Pull request by cyx1231st**
Implements [A] and [C] in https://github.com/ceph/ceph/pull/27430

2-way 4K read/write
----

saturated connection (throughput):
`$perf_crimson_msgr --round=4194304 --cbs=4096 --sbs=4096 --depth=512 -c 3`

 run | 1st | 2nd | 3rd 
---|---|---|---
base line | 21.3406s | 21.4988s | 21.3055s
+gather buffer | 19.7826s | 19.7931s | 19.7753s
+increase out buffer size | 18.4555s | 18.5185s | 18.4318s

depth=1 (latency):
`$perf_crimson_msgr --round=1048576 --cbs=4096 --sbs=4096 --depth=1 -c 3`

 run | 1st | 2nd | 3rd 
---|---|---|---
reference point | 20.5859s | 20.5712s | 20.3843s
+gather buffer | 21.0238s | 21.0726s | 21.0982s
+increase out buffer size | 20.8363s | 20.9622s | 21.1441s

2-way 1M read/write
----

saturated connection (throughput):
`$perf_crimson_msgr --round=65536 --cbs=1048576 --sbs=1048576 --depth=512 -c 3`

 run | 1st | 2nd | 3rd 
---|---|---|---
reference point | 25.1085s | 24.8107s | 25.0559s
+gather buffer | 24.6698s | 24.7343s | 24.5978s
+increase out buffer size | 24.7577s | 24.8432s | 24.8043s

depth=1 (latency):
`$perf_crimson_msgr --round=65536 --cbs=1048576 --sbs=1048576 --depth=1 -c 3 `

 run | 1st | 2nd | 3rd 
---|---|---|---
 reference point | 30.9145s | 30.8507s | 30.8499
 +gather buffer | 30.907s | 30.8757s | 30.8697s
 +increase out buffer size | 31.0175s | 30.9355s | 30.8921s

### CODE REVIEW

File: src/crimson/net/ProtocolV1.h [hunk 1]
```diff
@@ -26,11 +26,11 @@ class ProtocolV1 final : public Protocol {
...
-  seastar::future<> write_message(MessageRef msg) override;
-
-  seastar::future<> do_keepalive() override;
-
-  seastar::future<> do_keepalive_ack() override;
+  ceph::bufferlist do_sweep_messages(
+      const std::deque<MessageRef>& msgs,
+      size_t num_msgs,
+      bool require_keepalive,
+      bool require_keepalive_ack) override;
```
Comment (tchaikov) [inline src/crimson/net/ProtocolV1.h]: might want to take this opportunity to use `final`.

File: src/crimson/net/ProtocolV2.cc [hunk 2]
```diff
@@ -1332,46 +1332,54 @@ seastar::future<> ProtocolV2::send_reconnect_ok()
...
+                             header.reserved};
+
+    auto message = MessageFrame::Encode(header2,
+        msg->get_payload(), msg->get_middle(), msg->get_data());
+    logger().debug("{} write msg type={} off={} seq={}",
+                   conn, header2.type, header2.data_off, header2.seq);
+    bl.append(message.get_buffer(session_stream_handlers));
+  });
+
+  return std::move(bl);
```
Comment (tchaikov) [inline src/crimson/net/ProtocolV2.cc]: see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-return-move-local

File: src/crimson/net/ProtocolV2.h [hunk 3]
```diff
@@ -25,11 +25,11 @@ class ProtocolV2 final : public Protocol {
...
-  seastar::future<> write_message(MessageRef msg) override;
-
-  seastar::future<> do_keepalive() override;
-
-  seastar::future<> do_keepalive_ack() override;
+  ceph::bufferlist do_sweep_messages(
+      const std::deque<MessageRef>& msgs,
+      size_t num_msgs,
+      bool require_keepalive,
+      bool require_keepalive_ack) override;
```
Comment (tchaikov) [inline src/crimson/net/ProtocolV2.h]: `final` if you please.

File: src/crimson/net/Protocol.h [hunk 4]
```diff
@@ -61,6 +57,17 @@ class Protocol {
   seastar::gate pending_dispatch;
   AuthConnectionMetaRef auth_meta;
 
+ private:
+  bool closed = false;
+  // become valid only after closed == true
+  seastar::shared_future<> close_ready;
+
+// TODO: encapsulate as a state-machine for write
```
Comment (tchaikov) [inline src/crimson/net/Protocol.h]: what is this commit for? could you move it into its own PR where the state machine is implemented?

File: src/crimson/net/Socket.h [hunk 5]
```diff
@@ -35,7 +35,7 @@ class Socket
     : sid{seastar::engine().cpu_id()},
       socket(std::move(_socket)),
       in(socket.input()),
-      out(socket.output()) {}
+      out(socket.output(65536)) {}
```
Comment (tchaikov) [inline src/crimson/net/Socket.h]: magic number. could you add a very short comment to explain why we need to set it?

File: src/crimson/net/ProtocolV1.h (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/net/ProtocolV1.h]: I thought the class `ProtocolV1` was already marked `final`.
Please see discussion https://github.com/ceph/ceph/pull/26710#discussion_r270933077

File: src/crimson/net/ProtocolV2.h (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/net/ProtocolV2.h]: I thought the class `ProtocolV2` was already marked `final`.
Please see discussion https://github.com/ceph/ceph/pull/26710#discussion_r270933077

File: src/crimson/net/Socket.h (hunk 5 above)
Comment (cyx1231st) [reply src/crimson/net/Socket.h]: Comment added.

File: src/crimson/net/ProtocolV2.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/net/ProtocolV2.cc]: Yep, this is pessimization, fixed.

File: src/crimson/net/Protocol.h (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/net/Protocol.h]: It is already a state-machine now, and I just group them together here.
https://github.com/ceph/ceph/pull/27788/commits/04969f81508238f5e0839ffbbb86dcbd98746d19 should be more clear.

I'm just wondering if we can make the `Protocol` class much more lightweight, containing only the protocol-level differences in handshake logic and buffer-layouts. Other parts like this can be moved out to `SocketConnection`, in the future.

File: src/crimson/net/ProtocolV2.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/net/ProtocolV2.cc]: did you forget to push the updated change?

File: src/crimson/net/Socket.h (hunk 5 above)
Comment (tchaikov) [reply src/crimson/net/Socket.h]: i failed to find the comment =(

File: src/crimson/net/ProtocolV2.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/net/ProtocolV2.cc]: Sorry, seems I left them in another branch when doing tests, now they are included.

File: src/crimson/net/Socket.h (hunk 5 above)
Comment (cyx1231st) [reply src/crimson/net/Socket.h]: Sorry, seems I left them in another branch when doing tests, now they are included.

File: src/crimson/net/SocketConnection.h [hunk 6]
```diff
@@ -58,9 +58,9 @@ class SocketConnection : public Connection {
   bool update_rx_seq(seq_num_t seq);
 
   // messages to be resent after connection gets reset
-  std::queue<MessageRef> out_q;
+  std::deque<MessageRef> out_q;
```
Comment (rzarzynski) [inline src/crimson/net/SocketConnection.h]: In the future we could take a look on Seastar's `chunked_fifo` for that.

Comment (tchaikov) [reply src/crimson/net/SocketConnection.h]: logged as https://github.com/ceph/ceph/projects/2#card-20857269
